base/strings/util.jl

```diff
@@ -388,13 +388,83 @@ _replace(io, repl, str, r, pattern) = print(io, repl)
 _replace(io, repl::Function, str, r, pattern) =
     print(io, repl(SubString(str, first(r), last(r))))
 
-replace(str::String, pat_repl::Pair{Char}; count::Integer=typemax(Int)) =
-    replace(str, equalto(first(pat_repl)) => last(pat_repl); count=count)
-replace(str::String, pat_repl::Pair{<:Union{Tuple{Vararg{Char}},AbstractVector{Char},Set{Char}}};
-        count::Integer=typemax(Int)) =
-    replace(str, occursin(first(pat_repl)) => last(pat_repl), count)
+# replace Char with function that compares with this Char
+pairmap(p::Pair{Char}) = equalto(p.first) => p.second
+
+# replace collection of Char with function that checks occurrence in this collection
+function pairmap(p::Pair{<:Union{Tuple{Vararg{Char}},AbstractVector{Char},Set{Char}}})
+    occursin(p.first) => p.second
+end
+pairmap(p::Pair) = p
+
+# find the first of a list of patterns.
+# if several match at the same start position prefer the longer one
```

Comment:
Wouldn't it be better to just retain the first matching pair? At least that sounds simpler to explain. Or is that to match the behavior of other implementations?

Reply:
I think, this is special for strings. In case of ties, I do not simply take the first match, give precedence to  the longer match, because it is more specific. Maybe the example  demonstrates its usefulness:
```
replace("one house, two houses", "house"=>"Haus", "houses"=>"Häuser", "one"=>"ein", "two"=>"zwei")
```

Reply:
But can't the user just choose the appropriate order for the pairs? That won't work for regular expressions (since the length of the match may not be known in advance), but people would better use a single complex regular expression instead I think.

Another potential issue with not simply following the order in which pairs are provided is performance. One of the motivations for allowing for multiple pairs was to be more efficient than calling `replace` multiple times with different pairs. Doesn't having to match all pairs and compare them make the function slower? Or is that cost only paid when patterns overlap?

Reply:
The cost may increase only, if shorter patterns, which match are followed by longer patterns, which match at the same position. In other cases, I can't know in advance, it a later pattern does not match at an earlier position. 
BTW I would prefer to use a modified version of `findnext`, which has an extra parameter `does_not_start_after`, which would allow not to search the full string for all patterns, but stop earlier for the later patterns, if there was a match before.

Reply:
The docu text for `replace` has been updated. That should make the rule clear to the user of teh new feature:
```
510 In the case of multiple patterns, the precedence is by low start position of matched pattern,
511 then longer pattern, then order of the input list.
```

Reply:
What do you think about the additional optional argument for `findnext(,::AbstractString,...)` ?
I think the `does_not_start_after` could reduce search-times in a few cases. It would have another effect, than simply replacing the input string by a shortened substring.
The contract of this modified `findnext(pattern, input, start, not_after` would sound like:
"Give me the position of the next pattern match after `start`, but don't try input beyond `not_after` "